File: android/app/src/main/kotlin/com/stillmoment/infrastructure/audio/AttunementResolver.kt
package com.stillmoment.infrastructure.audio

import com.stillmoment.domain.models.Attunement
import com.stillmoment.domain.models.CustomAudioType
import com.stillmoment.domain.models.ResolvedAttunement
import com.stillmoment.domain.repositories.CustomAudioRepository
import com.stillmoment.domain.services.AttunementResolverProtocol
import javax.inject.Inject
import javax.inject.Singleton
import kotlinx.coroutines.runBlocking

/**
 * Infrastructure implementation of AttunementResolverProtocol.
 *
 * Transparently resolves attunement IDs by checking built-in attunements first
 * (filtered by current language), then falling back to user-imported custom attunements.
 *
 * Uses [runBlocking] internally because the resolver protocol is synchronous
 * (called from pure reducer functions that cannot suspend).
 */
@Singleton
class AttunementResolver @Inject constructor(
    private val customAudioRepository: CustomAudioRepository
) : AttunementResolverProtocol {

    override fun resolve(id: String): ResolvedAttunement? {
        // Try built-in attunement first (language-filtered)
        val attunement = Attunement.find(id)
        if (attunement != null && attunement.availableLanguages.contains(Attunement.currentLanguage)) {
            return ResolvedAttunement(
                id = attunement.id,
                displayName = attunement.localizedName,
                durationSeconds = attunement.durationSeconds
            )
        }

        // Try custom attunement
        val customFile = runBlocking { customAudioRepository.findFile(id) }
        if (customFile != null && customFile.type == CustomAudioType.ATTUNEMENT) {
            return ResolvedAttunement(
                id = customFile.id,
                displayName = customFile.name,
                durationSeconds = customFile.durationMs?.let { (it / 1000).toInt() } ?: 0
            )
        }

        return null
    }

    override fun allAvailable(): List<ResolvedAttunement> {
        // Built-in attunements for current language
        val builtIn = Attunement.availableForCurrentLanguage().map { item ->
            ResolvedAttunement(
                id = item.id,
                displayName = item.localizedName,
                durationSeconds = item.durationSeconds
            )
        }

        // Custom attunements
        val custom = runBlocking {
            customAudioRepository.loadAll(CustomAudioType.ATTUNEMENT)
        }.map { file ->
            ResolvedAttunement(
                id = file.id,
                displayName = file.name,
                durationSeconds = file.durationMs?.let { (it / 1000).toInt() } ?: 0
            )
        }

        return builtIn + custom
    }
}

File: ios/StillMoment/Infrastructure/Services/AttunementResolver.swift
//
//  AttunementResolver.swift
//  Still Moment
//
//  Infrastructure - Unified attunement resolution (built-in + custom)
//

import Foundation
import OSLog

/// Resolves attunement IDs by checking built-in attunements and custom audio repository.
///
/// Built-in attunements are language-filtered (only available for the device language).
/// Custom attunements are always available regardless of language.
final class AttunementResolver: AttunementResolverProtocol {
    private let customAudioRepository: CustomAudioRepositoryProtocol

    init(customAudioRepository: CustomAudioRepositoryProtocol) {
        self.customAudioRepository = customAudioRepository
    }

    func resolve(id: String) -> ResolvedAttunement? {
        // Try built-in attunement first
        if let attunement = Attunement.find(byId: id),
           attunement.availableLanguages.contains(Attunement.currentLanguage) {
            return ResolvedAttunement(
                id: attunement.id,
                displayName: attunement.name,
                durationSeconds: attunement.durationSeconds(for: Attunement.currentLanguage)
            )
        }

        // Try custom attunement (UUID-based ID)
        if let uuid = UUID(uuidString: id),
           let customFile = self.customAudioRepository.findFile(byId: uuid),
           customFile.type == .attunement {
            return ResolvedAttunement(
                id: id,
                displayName: customFile.name,
                durationSeconds: customFile.duration.map { Int($0) } ?? 0
            )
        }

        return nil
    }

    func resolveAudioURL(id: String) throws -> URL {
        // Try custom attunement first (UUID-based)
        if let uuid = UUID(uuidString: id),
           let customFile = self.customAudioRepository.findFile(byId: uuid),
           let url = self.customAudioRepository.fileURL(for: customFile) {
            return url
        }

        // Try built-in attunement
        guard let filename = Attunement.audioFilenameForCurrentLanguage(id) else {
            Logger.audio.error(
                "Attunement not found or not available",
                metadata: ["id": id]
            )
            throw AudioServiceError.soundFileNotFound
        }

        let components = filename.split(separator: ".", maxSplits: 1)
        let name = String(components.first ?? "")
        let ext = components.count > 1 ? String(components.last ?? "") : ""

        guard let url = Bundle.main.url(
            forResource: name,
            withExtension: ext,
            subdirectory: "IntroductionAudio"
        ) else {
            Logger.audio.error(
                "Attunement audio file not found in bundle",
                metadata: ["filename": filename]
            )
            throw AudioServiceError.soundFileNotFound
        }
        return url
    }

    func allAvailable() -> [ResolvedAttunement] {
        // Built-in attunements for current language
        var result = Attunement.availableForCurrentLanguage().map { attunement in
            ResolvedAttunement(
                id: attunement.id,
                displayName: attunement.name,
                durationSeconds: attunement.durationSeconds(for: Attunement.currentLanguage)
            )
        }

        // Custom attunements
        let customFiles = self.customAudioRepository.loadAll(type: .attunement)
        result += customFiles.map { file in
            ResolvedAttunement(
                id: file.id.uuidString,
                displayName: file.name,
                durationSeconds: file.duration.map { Int($0) } ?? 0
            )
        }

        return result
    }
}
